### PR TITLE
Segregate data in symbolic rewriting

### PR DESCRIPTION
**Pull request by cos**
**please review**

### CODE REVIEW

**Comment by cos** [comment]
Separates Data from rewriting information for ConstrainedTerm, SymbolicConstraint, and SymbolicUnifier.

Please review.

File: src/javasources/KTool/src/org/kframework/backend/java/kil/ConstrainedTerm.java [hunk 1]
```diff
@@ -32,27 +32,79 @@
...
 
-    /**
-     * Represents key lookups of builtin data-structures as a symbolic
-     * constraint.
-     */
-    private final SymbolicConstraint lookups;
-    private final SymbolicConstraint constraint;
     private final TermContext context;
 
+    private SymbolicConstraint lookups;
```
Comment (dwightguth) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/ConstrainedTerm.java]: have you discussed the mutability issue with Andrei and Yilong?

Comment (cos) [inline src/javasources/KTool/src/org/kframework/backend/java/kil/ConstrainedTerm.java]: At a high level, I've discussed extensively with Andrei and Yilong. You are right about those fields, they can be final. Still, lack of mutability annotations is a lesser worry in this code.

**Comment by cos** [comment]
@dwightguth, @andreistefanescu, @yilongli, please have a look.

**Comment by dwightguth** [comment]
I think there may be other places where a final annotation might be appropriate, but I'll leave this up to Andrei. I'd like him to review it before it is merged. It looks basically okay, but I don't know the code that well.

**Comment by cos** [comment]
I discussed with Andrei. While he said the changes look good, he advised me to delay the merge until he prepares a pull request of his current work.

**Pull request merged**